x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/search_after_bulk_create_suppressed_alerts.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SuppressedAlertService } from '@kbn/rule-registry-plugin/server';
+
+import { bulkCreateWithSuppression } from './bulk_create_with_suppression';
+import { addToSearchAfterReturn, getSuppressionMaxSignalsWarning } from './utils';
+import type {
+  SearchAfterAndBulkCreateParams,
+  SearchAfterAndBulkCreateReturnType,
+  WrapSuppressedHits,
+} from '../types';
+import { MAX_SIGNALS_SUPPRESSION_MULTIPLIER } from '../constants';
+
+import { createEnrichEventsFunction } from './enrichments';
+import { AlertSuppressionMissingFieldsStrategyEnum } from '../../../../../common/api/detection_engine/model/rule_schema';
+import type { AlertSuppressionCamel } from '../../../../../common/api/detection_engine/model/rule_schema';
+import { DEFAULT_SUPPRESSION_MISSING_FIELDS_STRATEGY } from '../../../../../common/detection_engine/constants';
+import { partitionMissingFieldsEvents } from './partition_missing_fields_events';
+interface SearchAfterAndBulkCreateSuppressedAlertsParams extends SearchAfterAndBulkCreateParams {
+  wrapSuppressedHits: WrapSuppressedHits;
+  alertTimestampOverride: Date | undefined;
+  alertWithSuppression: SuppressedAlertService;
+  alertSuppression?: AlertSuppressionCamel;
+}
+
+import type { SearchAfterAndBulkCreateFactoryParams } from './search_after_bulk_create_factory';
+import { searchAfterAndBulkCreateFactory } from './search_after_bulk_create_factory';
+
+/**
+ * search_after through documents and re-index using bulk endpoint
+ * and suppress alerts
+ */
+export const searchAfterAndBulkCreateSuppressedAlerts = async (
+  params: SearchAfterAndBulkCreateSuppressedAlertsParams
+): Promise<SearchAfterAndBulkCreateReturnType> => {
+  const {
+    wrapHits,
+    bulkCreate,
+    services,
+    buildReasonMessage,
+    ruleExecutionLogger,
+    tuple,
+    alertSuppression,
+    wrapSuppressedHits,
+    alertWithSuppression,
+    alertTimestampOverride,
+  } = params;
+
+  const bulkCreateExecutor: SearchAfterAndBulkCreateFactoryParams['bulkCreateExecutor'] = async ({
+    enrichedEvents,
+    toReturn,
+  }) => {
+    // max signals for suppression includes suppressed and created alerts
```

Comment:
I think this function could be more reusable if it just deals with the suppression part and lets the caller handle alert creation. The utility currently takes many parameters for `searchAfterAndBulkCreateFactory`, which is also called in the [search_after_bulk_create.ts](x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/search_after_bulk_create.ts).

By including only the suppression-specific parameters, we can use this utility in various executors. Additionally, certain parts of this utility's code can be shared with the search_after_bulk_create.

Reply:
Suppression logic is tightly interconnected with suppression.
In this executor `bulkCreate` is called to create non suppressed alerts with missing fields, and alerts are suppressed and created in `bulkCreateWithSuppression`. Then there is a logic specific to alert truncation, that can't be handled out of scope of alert creation.
So, code in this executor is specific for alert suppression only.  All common logic between `searchAfterAndBulkCreateSuppressedAlerts` and `searchAfterAndBulkCreate` is implemented within `searchAfterAndBulkCreateFactory`. While, suppression or non-suppression alert creation handled in corresponding executor.

In future, executor might be extracted as a separate utility, when it will be needed only for suppression/creation, without exhaustive search. For example, for EQL case

Reply:
@WafaaNasr it sounds like this could be very helpful to allow us to better unit test and resuse the functionality. Could we create a ticket to track this improvement and perhaps include it with the EQL alert suppression work?

Reply:
Thanks Vitalii and Yara!
 We talked about this matter and agreed that Vitalli will focus on extracting the [logic](https://github.com/elastic/kibana/pull/174241/files/55dde081b32b47d5200558fe1f2e7ef23a656ced#diff-4a7b759089d7881df21f2b6e881f3c27cc30c1016dfd67a15198388c8e1b5bb5R54-R106) here to make it reusable for other rule types in a different PR